Recording of drawing commands for undo/redo

dov opened the pull request:
This branch adds support for undo/redo as follows:

1. Added the class `DrawingCommand` that store either a pen or a erase action with all associated parameters. The coordinates are stored in relative window size coordinates.
2. Added the class `DrawingCommandList` that stores a list of drawing commands as well as a redo buffer. The undo command pushes the drawing commands from the drawing command list to redo command list. And the redo command does vice versa. The drawing command can also paint itself in a `Cairo.Surface`.
3. Changed `Storage` so that instead of storing images it stores a `DrawingCommandList` per slice. 
4. Change `Drawing` to use the DrawingCommandList to paint itself on undo, redo, or slide change.

fnevgeny commented:
For me, it doesn't compile:

`src/classes/drawings/drawing_commands.vala:75.18-75.42: error: The name `is_empty' does not exist in the context of `GLib.List<pdfpc.Drawings.DrawingCommands.DrawingCommand?>' (glib-2.0)`

dov commented:
@fnevgeny Thanks. It turns out that `is_empty()` is a relative new feature of (s)list in vala. I replaced it with an alternate construct. Please let me know of any additional problems.

One thing that I have noticed that I haven't solved is that my undo/redo commands do not trigger a redraw on their own. The redraw is only done after e.g. moving the mouse. If you have any idea of how to solve that I'd very much appreciate it.

fnevgeny commented:
> @fnevgeny Thanks. It turns out that `is_empty()` is a relative new feature of (s)list in vala. I replaced it with an alternate construct.

Thanks, it's fine now.

> Please let me know of any additional problems.

I think the undo/redo actions should be disregarded when not in a drawing mode.

> One thing that I have noticed that I haven't solved is that my undo/redo commands do not trigger a redraw on their own. The redraw is only done after e.g. moving the mouse. If you have any idea of how to solve that I'd very much appreciate it.

Add `this.queue_pen_surface_draws();`

dov commented:
> > Please let me know of any additional problems.
> 
> I think the undo/redo actions should be disregarded when not in a drawing mode.

Sure, makes sense. I'll limit it to drawing or eraser mode.

> > One thing that I have noticed that I haven't solved is that my undo/redo commands do not trigger a redraw on their own. The redraw is only done after e.g. moving the mouse. If you have any idea of how to solve that I'd very much appreciate it.
> 
> Add `this.queue_pen_surface_draws();`

Great! I'll add it.

dov commented:
Ok, I carried out these two changes. Redrawing works now. :-)

fnevgeny commented:
> Ok, I carried out these two changes.

Hmm, I see no changes in the PR...

dov commented:
Sorry. I forgot to push last night. I just pushed the changes.

fnevgeny commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 .history
 .*.swp
 build/
+*~
```
Let's omit changes not related to the issue.

fnevgeny commented on src/classes/drawings/drawing.vala:
```diff
@@ -91,6 +91,8 @@ namespace pdfpc.Drawings {
         private int current_slide {get; set;}
         private Drawings.Storage.Base storage {get; protected set;}
 
+        private Drawings.DrawingCommands.DrawingCommandList drawing_command_list;
```
I guess there is no need to keep the the upper-level `Drawings` in the class hierarchy.

dov commented:
@fnevgeny Thanks for the comments. I just forced pushed an updated branch in which I did the following changes:

- Removed changes of .gitignore
- Removed the .Drawings class from the DrawingsCommands

fnevgeny commented:
>     * Removed the .Drawings class from the DrawingsCommands

There are other places like that, e.g., `pdfpc.DrawingCommands.` in storage.vala.

fnevgeny commented on src/classes/drawings/drawing_commands.vala:
```diff
@@ -0,0 +1,170 @@
+        double y1;
+        double x2;
+        double y2;
+        float lwidth;
```
Is there any reason to use SP floats instead of double? You anyway cast them later to doubles.

dov replied:
I was trying to adapt the type to its usage in order to save some space. But in practice it really doesn't matter, since you are not likely have more than a few thousand drawing commands per slide. I'll change all types it to double for consistency.

I just force-pushed again

fnevgeny commented:
Overall, please try using the same formatting as the rest of the code: parentheses around single-line if/else blocks, no spaces between function names and (arguments), 80 chars width (unless absolutely ugly), no trailing whitespaces,

dov commented:
Reformatted the code according to your suggestions. I hope it is ok now.

fnevgeny commented:
> Reformatted the code according to your suggestions. I hope it is ok now.

Uhm, not exactly. I guess it would be easier if you give me write access to your branch directly.

dov commented:
> > Reformatted the code according to your suggestions. I hope it is ok now.
> 
> Uhm, not exactly. I guess it would be easier if you give me write access to your branch directly.

Ok, I just gave you write access to the repository.

dov commented:
I found some more discrepancies between the conventions you described and the code that I fixed. Do you have a linter or a code formatter that you use for your convention, btw?

fnevgeny commented:
Oh no, you've made meantime some changes. I won't bother repeating the manual reformatting, please do it yourself then..

Also, a respective update is needed for the pdfpc man page (and perhaps pdfpc-demo.tex).

> I found some more discrepancies between the conventions you described and the code that I fixed.

Could be, the code base has seen many people working on it over the years, but lately we try to make it unified.

> Do you have a linter or a code formatter that you use for your convention, btw?

Not really, but we should :).

fnevgeny commented on src/classes/drawings/drawing.vala:
```diff
@@ -130,6 +134,23 @@ namespace pdfpc.Drawings {
+            // Add to the drawing command list
+            this.drawing_command_list.add_line(
+                tool.is_eraser,
+                x1,y1,x2,y2,
```
put spaces after the comas: x1, y1, etc

fnevgeny commented on src/classes/drawings/drawing_commands.vala:
```diff
@@ -0,0 +1,175 @@
+                if (is_eraser == last.is_eraser
+                    && Math.fabs(last.x2-x1)<epsilon
+                    && Math.fabs(last.y2-y1)<epsilon) {
+                    new_path =false;
```
put space after =

fnevgeny commented on src/classes/drawings/storage.vala:
```diff
@@ -61,31 +61,33 @@ namespace pdfpc.Drawings.Storage {
 
         public override void clear() {
-            storage = new Cairo.ImageSurface[this.metadata.get_slide_count()];
+            // This was the old comment. Not sure it makes sense...
```
Yes, it still makes sense. Not that I like this approach, though...

fnevgeny commented on src/classes/drawings/drawing_commands.vala:
```diff
@@ -0,0 +1,175 @@
+            drawing_commands.foreach((dc) => {
+                if (dc.is_eraser) {
+                    cr.set_operator(Cairo.Operator.CLEAR);
+                }
```
change to `} else {`

dov replied:
Thanks. I applied all your comment and pushed again.

fnevgeny commented:
BTW, for a very similar approach with history navigation I used `Gee.ArrayQueue` (see `history_bck` and `history_fwd`). And it has `clear()` :)

dov commented:
### 🥳
